vignettes/decorate-module-output.Rmd

```diff
@@ -0,0 +1,589 @@
+It is recommended to create a library of decorator functions that can be adapted to the specific object names used in `teal` modules.
+In the following example, focus on the `output_name` parameter to see how decorator can be applied to multiple modules:
+
+```{r}
+gg_xlab_decorator <- function(output_name) {
+  teal_transform_module(
+    ui = function(id) {
+      ns <- NS(id)
+      div(
+        textInput(ns("x_axis_title"), "X axis title", value = "x axis")
+      )
+    },
+    server = function(id, data) {
+      moduleServer(id, function(input, output, session) {
+        reactive({
+          req(data())
+          within(data(),
+            {
+              output_name <- output_name +
+                xlab(x_axis_title)
+            },
+            x_axis_title = input$x_axis_title,
+            output_name = as.name(output_name)
+          )
+        })
+      })
+    }
+  )
+}
+```
+
+Decorator failures are managed by an internal `teal` mechanism called **trigger on success**, which ensures that the `data`
+object within the module remains intact. If a decorator fails, it will be ignored, and an appropriate error message will be displayed.
+
+```{r}
+failing_decorator <- teal_transform_module(
+  ui = function(id) {
+    ns <- NS(id)
+    div(
+      textInput(ns("x_axis_title"), "X axis title", value = "x axis")
+    )
+  },
+  server = function(id, data) {
+    moduleServer(id, function(input, output, session) {
+      reactive(stop("This is error"))
+    })
+  }
+)
+```
+
+## Decorating Plots
+
+### Example Module
+
+To include decorators in a `teal` module, pass them as arguments (`ui_args` and `server_args`) to the module’s `ui` and 
+`server` components, where they will be used by `ui/srv_teal_transform_module`:
+
+
+```{r}
+tm_decorated_plot <- function(label = "module", transformators = list(), decorators = teal_transform_module()) {
+  module(
+    label = label,
+    ui = function(id, decorators) {
+      ns <- NS(id)
+      div(
+        selectInput(ns("dataname"), label = "select dataname", choices = NULL),
+        selectInput(ns("x"), label = "select x", choices = NULL),
+        selectInput(ns("y"), label = "select y", choices = NULL),
+        ui_teal_transform_data(ns("decorate"), transformators = decorators),
+        plotOutput(ns("plot")),
+        verbatimTextOutput(ns("text"))
+      )
+    },
+    server = function(id, data, decorators) {
+      moduleServer(id, function(input, output, session) {
+        observeEvent(data(), {
+          updateSelectInput(inputId = "dataname", choices = names(data()))
+        })
+
+        observeEvent(input$dataname, {
+          req(input$dataname)
+          updateSelectInput(inputId = "x", choices = colnames(data()[[input$dataname]]))
+          updateSelectInput(inputId = "y", choices = colnames(data()[[input$dataname]]))
+        })
+
+        plot_data <- debounce(
```

Comment:
why `debounce`?

Reply:
No idea to be honest

Reply:
Would be great to get rid of this. 

We should aim to have a very simple examples so that the reader can keep focus on what's being described and not on anything extra - such as this one. Also - this gives some false impression that the usage of the `debuonce()` is required for this functionality which is not true.

Reply:
@m7pr It happens probably because of the reactive which triggers on change of data, updated `input$dataname` and updated `input$x` and `input$y`. Problem here is that any `updateSelectInput` triggers a new reactive cycle.

<img width="539" alt="image" src="https://github.com/user-attachments/assets/bef65955-0aed-45ab-84bb-ab9a482d155d">

Alternative solution could be to `isolate(data())` and `isolate(input$dataname)`, but this would be controversial solution here as ggplot should be recomputed when `data()` changes, when `dataname` changes, and then `x` or `y` changes.

Reply:
I found another aspect of this "extra-reactivity" with the sample app at the bottom. This seems more problematic.

Where modifying the input of another module triggers the decorator's expression of the first module. Illustrated by the circle emojis 🔴 (with decorator) and 🔵 (without decorator)

[Screencast from 2024-11-18 14-12-48.webm](https://github.com/user-attachments/assets/db6445be-71b6-4aad-8136-a1fab8d8cbc6)

The messages inside the mode require the following code added in `example_module`

```r
        # Added to example_module() just before table_data_decorated
        observeEvent(table_data(), logger::log_fatal("table_data reactive {label}"))
        observeEvent(input$dataname, logger::log_fatal("input$dataname reactive {label}"))
        observeEvent(input, logger::log_fatal("input reactive {label}"))
```

<details>

<summary>App code</summary>

```r
pkgload::load_all("teal")

options(teal.log_level = "INFO")

output_decorator <- teal_transform_module(
  label = "a label",
  server = make_teal_transform_server(
    expression({
      logger::log_warn("Expression being called! :: ", paste(ls(), collapse = ", "))
      object <- rev(object)
    })
  )
)

init(
  data = teal.data::teal_data(data1 = iris, data2 = mtcars, data3 = chickwts),
  modules = modules(
    example_module("module 🔴", decorators = output_decorator),
    example_module("module 🔵 (no decorator)", datanames = c("data2", "data3"))
  )
) |> shiny::runApp()
```

</details>

Reply:
@averissimo also found the same in example_module https://github.com/insightsengineering/teal/pull/1357#discussion_r1846519344

Reply:
The problem I reported is fixed after 5b402e268 and was unrelated.

The `debounce` issue can be solved by checking if `x` and `y` exist as colnames of the dataname. 

<details>

<summary>Suggestion here</summary>

```diff
diff --git a/vignettes/decorate-module-output.Rmd b/vignettes/decorate-module-output.Rmd
index 667c8a994..736d4991d 100644
--- a/vignettes/decorate-module-output.Rmd
+++ b/vignettes/decorate-module-output.Rmd
@@ -234,22 +234,29 @@ tm_decorated_plot <- function(label = "module", transformators = list(), decorat
           updateSelectInput(inputId = "x", choices = colnames(data()[[input$dataname]]))
           updateSelectInput(inputId = "y", choices = colnames(data()[[input$dataname]]))
         })
+        
+        dataname <- reactive(req(input$dataname))
+        x <- reactive({
+          req(input$x, input$x %in% colnames(data()[[dataname()]]))
+          input$x
+        })
+        
+        y <- reactive({
+          req(input$y, input$y %in% colnames(data()[[dataname()]]))
+          input$y
+        })
 
-        plot_data <- debounce(
-          reactive({
-            req(input$dataname, input$x, input$y)
-            within(data(),
-              {
-                plot <- ggplot2::ggplot(dataname, ggplot2::aes(x = x, y = y)) +
-                  ggplot2::geom_point()
-              },
-              dataname = as.name(input$dataname),
-              x = as.name(input$x),
-              y = as.name(input$y)
-            )
-          }), 200
-        )
-
+        plot_data <- reactive({
+          req(dataname(), x(), y())
+          
+          within(data(),
+            plot <- ggplot2::ggplot(dataname, ggplot2::aes(x = x, y = y)) + ggplot2::geom_point(),
+            dataname = req(as.name(input$dataname)),
+            x = as.name(x()),
+            y = as.name(y())
+          )
+        })
+        
```

</details>

Reply:
Thanks, @averissimo I think this is still simple enough! For sure less clutter than this obscure debounce. I'll update the vignette's code for all our examples (single decorator case and N decorator case).

Reply:
I updated the vignette with the following suggestion. Pushed the change in here 0003fef718eb10310f3a8be8d95c1db86ec02d8e

Reply:
Thanks both! This is much better now